Add CSV backend

kyleam opened the pull request:
This has a couple of advantages over the Text backend.
1. The values are stored as in a table per chain, not a file per
   variable.  This is easier to inspect and work with directly if
   desired (e.g., with pd.read_csv).
2. Values are stored during sampling, not kept in memory.

This relies on Pandas, which currently isn't listed as a hard
dependency.

kyleam committed:
Add CSV backend

This has a couple of advantages over the Text backend.

1. The values are stored as in a table per chain, not a file per
   variable.  This is easier to inspect and work with directly if
   desired (e.g., with pd.read_csv).

2. Values are stored during sampling, not kept in memory.

twiecki commented:
This is great. Do you think we could use odo to abstract away from the backend? https://odo.readthedocs.org/en/latest/

kyleam commented:
I started reading up on and experimenting with blaze, odo, and friends
this weekend, and they all look really nice.  But at I don't yet have a
clear picture of how to best integrate them and would be glad to see any
suggestions or proposals for this (but perhaps we should start a
separate, dedicated issue for exploring that?).

datnamer commented:
@kyleam, I think dask with its array of parallel shared memory, async and distributed schedulers and chunking system would be the best bet, aside from odo, for integration. More info, including an offer from the library author to help,  can be found in #707. Matt can help with use case guidance as well.

jsalvatier commented on pymc3/tests/test_csv_backend.py:
```diff
@@ -0,0 +1,114 @@
+def test_create_flat_names_1d():
+    shape = 2,
+    result = csv._create_flat_names('x', shape)
+    expected = ['x__0', 'x__1']
```
Why two underscores?

kyleam replied:
Just to make the indices visually distinct since it's common for
variable names to have both underscores and digits.  But if people
have a strong preference against it, I'm ok changing it to one
underscore.

jsalvatier replied:
Makes sense. I think its good to keep.

jsalvatier commented:
Is pandas ever a hard dependency to install? Is there a reason not to make it a hard dep? 

Otherwise looks good to me.

kyleam commented:
I'm for making Pandas a non-optional dependency.  If we do that, I'd
be in favor of removing the Text backend completely and replacing it
with the CSV (perhaps renaming CSV to Text?).

jsalvatier commented:
@twiecki @fonnesbeck ? Thoughts on making pandas a hard dep?

fonnesbeck commented:
Given that Pandas is now a core library in the scientific stack, forcing users to install it is not unreasonable.

twiecki commented:
+1 on adding pandas as a hard dependency

jsalvatier commented:
@kyleam can you add the dependency to setup.py? Then I think we're good to go.

kyleam commented:
> can you add the dependency to setup.py?

Will do.

kyleam committed:
Replace Text backend with CSV

The main reason to keep the Text backend around was that pandas was an
optional dependency.  Now that this is no longer the case, the
original Text backend doesn't offer any advantages over the CSV
backend.  (Even if someone prefers not to write to files while
sampling, they can sample with the NDArray backend and then use the
CSV dump function.)

Rename CSV to Text.  This name is more appropriate because the values
being stored as plain text is the important feature, not which
delimiter is used.

kyleam committed:
Add pandas as hard dependency

kyleam commented:
Updated.  Does anyone have an issue with removing Text (and renaming CSV to Text)?

jsalvatier commented:
What additional capabilities beyond CSV does Text currently have? If it doesn't really have any, then I'm all for it.

kyleam commented:
None (see commit message).

jsalvatier commented:
Yeah, lets replace it then.
